In [7]:
# Import thư viện NLTK và download công cụng NLTK
import nltk
nltk.download_shell()

NLTK Downloader
---------------------------------------------------------------------------
    d) Download   l) List    u) Update   c) Config   h) Help   q) Quit
---------------------------------------------------------------------------
Downloader> l

Packages:
  [ ] abc................. Australian Broadcasting Commission 2006
  [ ] alpino.............. Alpino Dutch Treebank
  [ ] averaged_perceptron_tagger Averaged Perceptron Tagger
  [ ] averaged_perceptron_tagger_eng Averaged Perceptron Tagger (JSON)
  [ ] averaged_perceptron_tagger_ru Averaged Perceptron Tagger (Russian)
  [ ] averaged_perceptron_tagger_rus Averaged Perceptron Tagger (Russian)
  [ ] basque_grammars..... Grammars for Basque
  [ ] bcp47............... BCP-47 Language Tags
  [ ] biocreative_ppi..... BioCreAtIvE (Critical Assessment of Information
                           Extraction Systems in Biology)
  [ ] bllip_wsj_no_aux.... BLLIP Parser: WSJ Model
  [ ] book_grammars....... Grammars from NLTK Book
  [ ] brown.

      Package gutenberg is already up-to-date!



---------------------------------------------------------------------------
    d) Download   l) List    u) Update   c) Config   h) Help   q) Quit
---------------------------------------------------------------------------
Downloader> q


In [21]:
nltk.download('gutenberg')
nltk.download('punkt')

[nltk_data] Downloading package gutenberg to /root/nltk_data...
[nltk_data]   Package gutenberg is already up-to-date!
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


True

In [15]:
# Sau khi hoàn tất, bạn có xem tên các tập tin có trong gói 'gutenberg' bằng lệnh fileids().
gb = nltk.corpus.gutenberg
print("Gutenberg files : ", gb.fileids())

Gutenberg files :  ['austen-emma.txt', 'austen-persuasion.txt', 'austen-sense.txt', 'bible-kjv.txt', 'blake-poems.txt', 'bryant-stories.txt', 'burgess-busterbrown.txt', 'carroll-alice.txt', 'chesterton-ball.txt', 'chesterton-brown.txt', 'chesterton-thursday.txt', 'edgeworth-parents.txt', 'melville-moby_dick.txt', 'milton-paradise.txt', 'shakespeare-caesar.txt', 'shakespeare-hamlet.txt', 'shakespeare-macbeth.txt', 'whitman-leaves.txt']


Để truy cập nội dung bên trong của một trong các tập tin này, trước tiên bạn chọn một tập tin, chẳng hạn Macbeth của Shakespeare (shakespeare-macbeth.txt), sử dụng hàm words(), rồi gán cho một biến nào đó.

In [16]:
macbeth = nltk.corpus.gutenberg.words('shakespeare-macbeth.txt')

In [17]:
len(macbeth) # xem chiều dài văn bản

23140

In [18]:
macbeth[:10] #Hiển thị 10 từ đầu tiên

['[',
 'The',
 'Tragedie',
 'of',
 'Macbeth',
 'by',
 'William',
 'Shakespeare',
 '1603',
 ']']

Hàm trên trích ra 10 từ đầu tiên là tiêu đề của tập tin, dấu ngoặc vuông được tính là 1.
Ta muốn trích 5 câu đầu tiên của tập tin (một câu được kẹp trong cặp ngoặc vuông), ta dùng hàm sent()

In [22]:
macbeth_sents = nltk.corpus.gutenberg.sents('shakespeare-macbeth.txt')
print(macbeth_sents[:5])

[['[', 'The', 'Tragedie', 'of', 'Macbeth', 'by', 'William', 'Shakespeare', '1603', ']'], ['Actus', 'Primus', '.'], ['Scoena', 'Prima', '.'], ['Thunder', 'and', 'Lightning', '.'], ['Enter', 'three', 'Witches', '.']]


# **Tìm 1 từ với NLTK:**
Tìm từ ‘Stage’ xuất hiện trong văn bản text

In [23]:
text = nltk.Text(macbeth)
text.concordance('Stage')

Displaying 3 of 3 matches:
nts with Dishes and Seruice ouer the Stage . Then enter Macbeth Macb . If it we
with mans Act , Threatens his bloody Stage : byth ' Clock ' tis Day , And yet d
 struts and frets his houre vpon the Stage , And then is heard no more . It is 


Tìm từ xuất hiện trước và sau từ ‘Stage’

In [24]:
text.common_contexts(['Stage'])

the_. bloody_: the_,


Tìm từ tương tự từ ‘Stage’

In [25]:
text.similar('Stage')

day time face warre ayre king bleeding man reuolt serieant like
knowledge broyle shew head spring heeles hare thane skie


# **Phân tích tần số của các từ**
Phân tích số lần xuất hiện của từ:
Muốn xem 10 từ thông dụng nhất trong văn bản xuất hiện bao nhiêu lần, dùng lệnh \
most_common()

In [26]:
fd = nltk.FreqDist(macbeth)
fd.most_common(10)

[(',', 1962),
 ('.', 1235),
 ("'", 637),
 ('the', 531),
 (':', 477),
 ('and', 376),
 ('I', 333),
 ('of', 315),
 ('to', 311),
 ('?', 241)]

Stopword là những từ thông dụng, thường ít có ý nghĩa trong quá trình phân tích văn bản và thường cần được loại bỏ. \
Muốn download stopword

In [27]:
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


True

Muốn xem các stopword trong tiếng Anh, dùng lệnh

In [28]:
sw = set(nltk.corpus.stopwords.words('english'))
print(len(sw))
list(sw)[:10]

179


["isn't", 'don', 'above', 'just', 'such', 'mightn', 'of', 'haven', 'her', 't']

Có 179 stopword trong từ vựng tiếng Anh. Ta sẽ loại bỏ các từ stopword trong biến macbeth

In [29]:
macbeth_filtered = [w for w in macbeth if w.lower() not in sw]
len(macbeth_filtered)

14946

In [31]:
fd = nltk.FreqDist(macbeth_filtered)
fd.most_common(10)

[(',', 1962),
 ('.', 1235),
 ("'", 637),
 (':', 477),
 ('?', 241),
 ('Macb', 137),
 ('haue', 117),
 ('-', 100),
 ('Enter', 80),
 ('thou', 63)]

Bây giờ, 10 từ phổ biến nhất đầu tiên được trả về, các stopword đã được loại bỏ nhưng vẫn còn
các dấu câu, ta cần loại bỏ các dấu câu theo lệnh sau:

In [33]:
import string
punctuation = set(string.punctuation)
macbeth_filtered2 = [w.lower() for w in macbeth if w.lower() not in sw and w.lower() not in punctuation]
fd = nltk.FreqDist(macbeth_filtered2)
fd.most_common(10)

[('macb', 137),
 ('haue', 122),
 ('thou', 90),
 ('enter', 81),
 ('shall', 68),
 ('macbeth', 62),
 ('vpon', 62),
 ('thee', 61),
 ('macd', 58),
 ('vs', 57)]

# **Lựa chọn các từ trong văn bản**
Lựa chọn các từ trong văn bản:
Một dạng khác trong việc xử lý và phân tích dữ liệu là quá trình chọn một từ dựa vào 1 đặc điểm
nào đó. Ví dụ như bạn cần quan tâm rút trích từ dựa vào độ dài của từ. \
Rút trích các từ có độ dài lớn nhất, ví dụ các từ có độ dài lớn hơn 12 ký tự , dùng lệnh sau:


In [34]:
long_words = [w for w in macbeth if len(w)> 12]
sorted(long_words)

['Assassination',
 'Chamberlaines',
 'Distinguishes',
 'Gallowgrosses',
 'Metaphysicall',
 'Northumberland',
 'Voluptuousnesse',
 'commendations',
 'multitudinous',
 'supernaturall',
 'vnaccompanied']

Rút trích các từ có chứa chuỗi ‘ious’

In [35]:
ious_words = [w for w in macbeth if 'ious' in w]
ious_words = set(ious_words)
sorted(ious_words)

['Auaricious',
 'Gracious',
 'Industrious',
 'Iudicious',
 'Luxurious',
 'Malicious',
 'Obliuious',
 'Pious',
 'Rebellious',
 'compunctious',
 'furious',
 'gracious',
 'pernicious',
 'pernitious',
 'pious',
 'precious',
 'rebellious',
 'sacrilegious',
 'serious',
 'spacious',
 'tedious']

# **Bigrams và collocations**
Một phần cơ bản trong phân tích văn bản là xem xét các cặp từ thay thế cho các từ đơn. Từ ‘is’
và ‘yellow’ là 1 ví dụ về bigram vì chúng có thể kết hợp lại và có ý nghĩa, vì thế “is yellow” có
thể được tìm thấy trong văn bản. Ví dụ “fast food”, “pay attention”, “good morning”,..,những
bigram này gọi là collocation. \
Lọc các bigram sau khi đã loại các stopword và các dấu câu, dùng lệnh sau:

In [36]:
bgrms = nltk.FreqDist(nltk.bigrams(macbeth_filtered2))
bgrms.most_common(15)

[(('enter', 'macbeth'), 16),
 (('exeunt', 'scena'), 15),
 (('thane', 'cawdor'), 13),
 (('knock', 'knock'), 10),
 (('st', 'thou'), 9),
 (('thou', 'art'), 9),
 (('lord', 'macb'), 9),
 (('haue', 'done'), 8),
 (('macb', 'haue'), 8),
 (('good', 'lord'), 8),
 (('let', 'vs'), 7),
 (('enter', 'lady'), 7),
 (('wee', 'l'), 7),
 (('would', 'st'), 6),
 (('macbeth', 'macb'), 6)]

Ngoài bigram ra, còn có trigram, sự kết hợp của 3 từ, ta dùng lệnh trigrams()

In [37]:
tgrms = nltk.FreqDist(nltk.trigrams (macbeth_filtered2))
tgrms.most_common(10)

[(('knock', 'knock', 'knock'), 6),
 (('enter', 'macbeth', 'macb'), 5),
 (('enter', 'three', 'witches'), 4),
 (('exeunt', 'scena', 'secunda'), 4),
 (('good', 'lord', 'macb'), 4),
 (('three', 'witches', '1'), 3),
 (('exeunt', 'scena', 'tertia'), 3),
 (('thunder', 'enter', 'three'), 3),
 (('exeunt', 'scena', 'quarta'), 3),
 (('scena', 'prima', 'enter'), 3)]

# **Sử dụng văn bản trên mạng**
Các ví dụ trước, chúng ta đã làm việc với văn bản sử dụng thư viện NLTK và gói ‘gutenberg’.
Nhưng trong thực tế, bạn sẽ cần phải truy cập Internet để trích xuất văn bản và thu thập nó như
một kho tài liệu được sử dụng để phân tích với NLTK. \
Import thư viện và mở url để đọc file

In [39]:
from urllib import request
url = "http://www.gutenberg.org/files/2554/2554-0.txt"
response = request.urlopen(url)
raw = response.read().decode('utf8')
raw[:75]

'\ufeffThe Project Gutenberg eBook of Crime and Punishment, by Fyodor Dostoevsky\r'

Thay bằng các lệnh sau:

In [44]:
from urllib import request
url = "http://www.gutenberg.org/files/2554/2554-0.txt"
response = request.urlopen(url)
raw = response.read().decode('utf-8')
# Xóa BOM nếu nó tồn tại
if raw.startswith('\ufeff'):
    raw = raw[1:]
print(raw[:75])

The Project Gutenberg eBook of Crime and Punishment, by Fyodor Dostoevsky



Thực hiện các lệnh sau:

In [45]:
tokens = nltk.word_tokenize (raw)
webtext = nltk.Text (tokens)
webtext[:12]

['The',
 'Project',
 'Gutenberg',
 'eBook',
 'of',
 'Crime',
 'and',
 'Punishment',
 ',',
 'by',
 'Fyodor',
 'Dostoevsky']

# **Rút trích văn bản từ trang html**
Trong ví dụ vừa rồi, ta tạo 1 copus NLTK từ 1 văn bản download từ Internet. Nhưng hầu như các
văn bản trên Internet ở dạng trang html. Trong phần này, chúng ta sẽ tìm hiểu cách rút trích văn
bản từ trang html.

In [46]:
url = "http://news.bbc.co.uk/2/hi/health/2284783.stm"
html = request.urlopen(url).read().decode('utf8')
html[:120]

'<!doctype html public "-//W3C//DTD HTML 4.0 Transitional//EN" "http://www.w3.org/TR/REC-html40/loose.dtd">\r\n<html>\r\n<hea'

Ta dùng thư viện bs4 (BeautifulSoup) cung cấp cho bạn các trình phân tích cú pháp phù hợp có
thể nhận dạng HTML và trích xuất văn bản.

In [47]:
from bs4 import BeautifulSoup
raw = BeautifulSoup(html, "lxml").get_text()
tokens = nltk.word_tokenize(raw)
text = nltk.Text(tokens)

# **Phân tích cảm xúc người dùng**
Nhờ nhận xét và đánh giá của người dùng, các thuật toán phân tích cảm xúc có thể đánh giá mức
độ đánh giá hoặc đánh giá dựa trên các từ khóa nhất định. Mức độ đánh giá được gọi là ý kiến và
có ba giá trị có thể có: positive, neutral hoặc negative. Việc đánh giá ý kiến này do đó trở thành
một dạng phân loại. \
Mục đích của ví dụ dưới đây là tìm các từ lặp lại nhiều nhất dạng positive và dạng negative hoặc
các từ khóa liên quan đến một ý kiến nào đó, sử dụng thuật toán Naïve Bayes để phân loại cảm
xúc người dùng dựa vào các đoạn văn bản movie_reviews của họ. \
Trước tiên, download các movie review dùng lệnh sau:


In [48]:
nltk.download('movie_reviews')

[nltk_data] Downloading package movie_reviews to /root/nltk_data...
[nltk_data]   Unzipping corpora/movie_reviews.zip.


True

Sau đó xây dựng tập train dựa vào copus trên. Tạo 1 mảng tên là documents. Mảng này chứa cột
đầu tiên là nội dung review của người dùng, và cột thứ 2 là cột đánh giá positive, neutral, or
negative. Sau đó trộn các dòng này ngẫn nhiên

In [49]:
import random
reviews = nltk.corpus.movie_reviews
documents = [(list(reviews.words(fileid)), category)
for category in reviews.categories()
for fileid in reviews.fileids(category)]
random.shuffle(documents)

Xem nội dung review đầu tiên (dòng 0, cột 0)

In [50]:
first_review = ' '.join(documents[0][0])
print(first_review)

it ' s a fact that a good thriller or action movie doesn ' t need violence to be good or worth watching . all it takes is potential violence to make the audience bite their nails . and what kind of violence could be more efficient than a global thermo - nuclear war ? and potential violence is the premiss wargames is built upon . a computer whiz - kid , david , is usually contended with hacking into the school computer to change his grades . but after having read an advertisment for an upcoming computer game , he wants to be the first to play it . instead of getting into the computer at protovision software he accidentally comes to the frontgate of norad . with the help of jennifer , a girl in his biology class , he starts to play a nice game of nuclear war with wopr , war operation plan response ( a top - notch computer at the time , but now hardly faster than my own computer ) i remember being 12 and watching this at the cinema . it was very efficent at giving you a scare in those day

Xem kết quả review đầu tiên (dòng 0, cột 1)

In [51]:
documents[0][1]

'pos'

Ta cần tạo bảng phân phối tần số các từ trong copus, bảng này cần chuyển sang dạng list, ta dùng
hàm list()

In [52]:
all_words = nltk.FreqDist(w.lower() for w in reviews.words())
word_features = list(all_words)

Sau đó, bước tiếp theo là xác định một hàm để tính toán các đặc trưng, tức là những từ đủ quan
trọng để thiết lập ý kiến của một review

In [53]:
def document_features(document, word_features):
  document_words = set(document)
  features = {}
  for word in word_features:
    features['{}'.format(word)] = (word in document_words)
  return features

Khi bạn định nghĩa hàm document_features(), bạn tạo 1 tập các documents

In [55]:
featuresets = [(document_features(d,word_features), c) for (d,c) in documents]
len(featuresets)

2000

Tạo tập train và tập test: 1500 dòng đầu dùng cho tập train và 500 dòng còn lại dùng cho tập test
để đánh giá độ chính xác của mô hình.

In [56]:
train_set, test_set = featuresets[1500:], featuresets[:500]
classifier = nltk.NaiveBayesClassifier.train(train_set)

Dùng thuật toán Naïve Bayes để phân loại, dùng thư viện NLTK. Sau đó tính toán độ chính xác
của thuật toán

In [57]:
train_set, est_set = featuresets[1500:], featuresets[:500]
classifier = nltk.NaiveBayesClassifier.train(train_set)
print(nltk.classify.accuracy(classifier, test_set))

0.744


Chúng ta đã hoàn tất việc phân tích, dưới đây các từ với trọng số lớn nhất của các review được
đánh giá là positive và negative

In [58]:
classifier.show_most_informative_features(10)

Most Informative Features
               portrayed = True              pos : neg    =     12.8 : 1.0
                anywhere = True              neg : pos    =     10.1 : 1.0
                  random = True              neg : pos    =     10.1 : 1.0
              nomination = True              pos : neg    =      9.8 : 1.0
             wonderfully = True              pos : neg    =      8.8 : 1.0
                  losing = True              pos : neg    =      8.6 : 1.0
                lifeless = True              neg : pos    =      8.6 : 1.0
               initially = True              pos : neg    =      8.0 : 1.0
                 luckily = True              pos : neg    =      8.0 : 1.0
                   awful = True              neg : pos    =      7.8 : 1.0


# **Bài tập ứng dụng**

1. Viết chương trình Python với thư viện NLTK để liệt kê các tên của copus

In [60]:
import nltk
from nltk.corpus import *
# List available corpora in NLTK
corpus_names = dir(nltk.corpus)
print("Danh sách các corpus:")
for name in corpus_names:
    print("- ", name)

Danh sách các corpus:
-  AlignedCorpusReader
-  AlpinoCorpusReader
-  BCP47CorpusReader
-  BNCCorpusReader
-  BracketParseCorpusReader
-  CHILDESCorpusReader
-  CMUDictCorpusReader
-  CategorizedBracketParseCorpusReader
-  CategorizedCorpusReader
-  CategorizedPlaintextCorpusReader
-  CategorizedSentencesCorpusReader
-  CategorizedTaggedCorpusReader
-  ChasenCorpusReader
-  ChunkedCorpusReader
-  ComparativeSentencesCorpusReader
-  ConllChunkCorpusReader
-  ConllCorpusReader
-  CorpusReader
-  CrubadanCorpusReader
-  DependencyCorpusReader
-  EuroparlCorpusReader
-  FramenetCorpusReader
-  IEERCorpusReader
-  IPIPANCorpusReader
-  IndianCorpusReader
-  KNBCorpusReader
-  LazyCorpusLoader
-  LinThesaurusCorpusReader
-  MTECorpusReader
-  MWAPPDBCorpusReader
-  MacMorphoCorpusReader
-  NKJPCorpusReader
-  NPSChatCorpusReader
-  NombankCorpusReader
-  NonbreakingPrefixesCorpusReader
-  OpinionLexiconCorpusReader
-  PPAttachmentCorpusReader
-  PanLexLiteCorpusReader
-  PanlexSwadeshCorpusR

2. Viết chương trình Python với thư viện NLTK để liệt kê danh sách các stopword bằng các
ngôn ngữ khác nhau.

In [62]:
# Liệt kê các ngôn ngữ có sẵn
print("Danh sách các ngôn ngữ có sẵn:")
for lang in nltk.corpus.stopwords.fileids():
    print("- ", lang)

# Chọn một ngôn ngữ (ví dụ: english)
lang = 'english'

# Liệt kê danh sách stopword cho ngôn ngữ đã chọn
print("\nDanh sách stopword cho english:")
for word in nltk.corpus.stopwords.words(lang):
    print(word)

Danh sách các ngôn ngữ có sẵn:
-  arabic
-  azerbaijani
-  basque
-  bengali
-  catalan
-  chinese
-  danish
-  dutch
-  english
-  finnish
-  french
-  german
-  greek
-  hebrew
-  hinglish
-  hungarian
-  indonesian
-  italian
-  kazakh
-  nepali
-  norwegian
-  portuguese
-  romanian
-  russian
-  slovene
-  spanish
-  swedish
-  tajik
-  turkish

Danh sách stopword cho english:
i
me
my
myself
we
our
ours
ourselves
you
you're
you've
you'll
you'd
your
yours
yourself
yourselves
he
him
his
himself
she
she's
her
hers
herself
it
it's
its
itself
they
them
their
theirs
themselves
what
which
who
whom
this
that
that'll
these
those
am
is
are
was
were
be
been
being
have
has
had
having
do
does
did
doing
a
an
the
and
but
if
or
because
as
until
while
of
at
by
for
with
about
against
between
into
through
during
before
after
above
below
to
from
up
down
in
out
on
off
over
under
again
further
then
once
here
there
when
where
why
how
all
any
both
each
few
more
most
other
some
such
no
nor
not
only
own
sa

3. Viết chương trình Python với thư viện NLTK để kiểm tra danh sách các stopword bằng các ngôn ngữ khác nhau

In [64]:
# Liệt kê các ngôn ngữ có sẵn
print("Danh sách các ngôn ngữ có sẵn:")
for lang in nltk.corpus.stopwords.fileids():
    print("- ", lang)

# Chọn một ngôn ngữ (ví dụ: english)
lang = 'english'

# Nhập từ cần kiểm tra
word_to_check = input("Nhập từ cần kiểm tra: ")

# Kiểm tra xem từ có phải là stopword không
if word_to_check.lower() in nltk.corpus.stopwords.words(lang):
    print(f"'{word_to_check}' là một stopword trong tiếng {lang}.")
else:
    print(f"'{word_to_check}' không phải là một stopword trong tiếng {lang}.")

Danh sách các ngôn ngữ có sẵn:
-  arabic
-  azerbaijani
-  basque
-  bengali
-  catalan
-  chinese
-  danish
-  dutch
-  english
-  finnish
-  french
-  german
-  greek
-  hebrew
-  hinglish
-  hungarian
-  indonesian
-  italian
-  kazakh
-  nepali
-  norwegian
-  portuguese
-  romanian
-  russian
-  slovene
-  spanish
-  swedish
-  tajik
-  turkish
Nhập từ cần kiểm tra: wouldn
'wouldn' là một stopword trong tiếng english.


4. Viết chương trình Python với thư viện NLTK để loại bỏ các stopword từ một văn bản đã cho.

In [69]:
import nltk
from nltk.corpus import stopwords
nltk.download('stopwords')

def remove_stopwords(text):
  stop_words = set(stopwords.words('english'))
  words = nltk.word_tokenize(text)
  filtered_words = [word for word in words if word.lower() not in stop_words]
  return ' '.join(filtered_words)

text = input('Nhập vào văn bản: ')
filtered_text = remove_stopwords(text)
print("\nVăn bản gốc:", text)
print("\nVăn bản sau khi loại bỏ stopword:", filtered_text)


[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


Nhập vào văn bản: The college will offer internships and hands-on practice in keeping giant pandas safe, breeding them, and keeping them from getting diseases. People think that the college will do a lot to protect the environment, restore animal areas, build a panda national park, help the panda population grow, and do research.

Văn bản gốc: The college will offer internships and hands-on practice in keeping giant pandas safe, breeding them, and keeping them from getting diseases. People think that the college will do a lot to protect the environment, restore animal areas, build a panda national park, help the panda population grow, and do research.

Văn bản sau khi loại bỏ stopword: college offer internships hands-on practice keeping giant pandas safe , breeding , keeping getting diseases . People think college lot protect environment , restore animal areas , build panda national park , help panda population grow , research .


5. Viết chương trình Python với thư viện NLTK bỏ qua các stopword từ danh sách các stopword.

In [71]:
def ignore_stopwords(text, ignored_words):
  stop_words = set(stopwords.words('english')) - set(ignored_words)
  words = nltk.word_tokenize(text)
  filtered_words = [word for word in words if word.lower() not in stop_words]
  return ' '.join(filtered_words)

text = input('Nhập vào văn bản: ')
ignored_words = input('\nNhập vào các stopword bạn muốn bỏ qua (phân cách bằng dấu phẩy): ').split(',')
filtered_text = ignore_stopwords(text, ignored_words)
print("\nVăn bản gốc:", text)
print("\nVăn bản sau khi bỏ qua các stopword:", filtered_text)


Nhập vào văn bản: The college will offer internships and hands-on practice in keeping giant pandas safe, breeding them, and keeping them from getting diseases. People think that the college will do a lot to protect the environment, restore animal areas, build a panda national park, help the panda population grow, and do research.

Nhập vào các stopword bạn muốn bỏ qua (phân cách bằng dấu phẩy): The

Văn bản gốc: The college will offer internships and hands-on practice in keeping giant pandas safe, breeding them, and keeping them from getting diseases. People think that the college will do a lot to protect the environment, restore animal areas, build a panda national park, help the panda population grow, and do research.

Văn bản sau khi bỏ qua các stopword: college offer internships hands-on practice keeping giant pandas safe , breeding , keeping getting diseases . People think college lot protect environment , restore animal areas , build panda national park , help panda population gr

6. Viết một chương trình Python với thư viện NLTK để tìm định nghĩa và ví dụ của một từ đã cho bằng WordNet từ Wikipedia.

In [74]:
!pip install wikipedia-api

In [77]:
import nltk
from nltk.corpus import wordnet as wn
import wikipediaapi

# Tải xuống dữ liệu NLTK cần thiết
nltk.download('wordnet')
nltk.download('omw-1.4')

def get_wordnet_info(word):
    synsets = wn.synsets(word)
    if not synsets:
        return None

    definitions = []
    examples = []

    for synset in synsets:
        definitions.append(synset.definition())
        examples.extend(synset.examples())

    return definitions, examples

def get_wikipedia_info(word):
    user_agent = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36"
    wiki_wiki = wikipediaapi.Wikipedia('en', headers={'User-Agent': user_agent})
    page = wiki_wiki.page(word)

    if not page.exists():
        return None

    return page.summary

def main(word):
    # Lấy định nghĩa và ví dụ về WordNet
    wordnet_info = get_wordnet_info(word)
    if wordnet_info:
        definitions, examples = wordnet_info
        print(f"Định nghĩa WordNet cho '{word}':")
        for i, definition in enumerate(definitions, 1):
            print(f"{i}. {definition}")

        print(f"\nVí dụ về WordNet cho '{word}':")
        for i, example in enumerate(examples, 1):
            print(f"{i}. {example}")
    else:
        print(f"Không tìm thấy thông tin WordNet cho '{word}'.")

    # Lấy bản tóm tắt Wikipedia
    wikipedia_info = get_wikipedia_info(word)
    if wikipedia_info:
        print(f"\nTóm tắt Wikipedia cho '{word}':")
        print(wikipedia_info)
    else:
        print(f"Không tìm thấy thông tin Wikipedia cho '{word}'.")

if __name__ == "__main__":
    word = input("\nNhập vào từ khóa: ")
    main(word)


[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package omw-1.4 to /root/nltk_data...
[nltk_data]   Package omw-1.4 is already up-to-date!



Nhập vào từ khóa: key
Định nghĩa WordNet cho 'key':
1. metal device shaped in such a way that when it is inserted into the appropriate lock the lock's mechanism can be rotated
2. something crucial for explaining
3. pitch of the voice
4. any of 24 major or minor diatonic scales that provide the tonal framework for a piece of music
5. a kilogram of a narcotic drug
6. a winged often one-seed indehiscent fruit as of the ash or elm or maple
7. United States lawyer and poet who wrote a poem after witnessing the British attack on Baltimore during the War of 1812; the poem was later set to music and entitled `The Star-Spangled Banner' (1779-1843)
8. a coral reef off the southern coast of Florida
9. (basketball) a space (including the foul line) in front of the basket at each end of a basketball court; usually painted a different color from the rest of the court
10. a list of answers to a test
11. a list of words or phrases that explain symbols or abbreviations
12. a generic term for any devic

7. Viết chương trình Python với thư viện NLTK để tìm tập hợp các từ đồng nghĩa và trái nghĩa
của một từ nào đó.

In [78]:
nltk.download('wordnet')

def find_synonyms_antonyms(word):
    synonyms = []
    antonyms = []
    for synset in wn.synsets(word):
        for lemma in synset.lemmas():
            synonyms.append(lemma.name())
            if lemma.antonyms():
                antonyms.append(lemma.antonyms()[0].name())
    return synonyms, antonyms

if __name__ == "__main__":
    word = input("Nhập vào từ khóa: ")
    synonyms, antonyms = find_synonyms_antonyms(word)
    print(f"Từ đồng nghĩa của '{word}': {set(synonyms)}")
    print(f"Từ trái nghĩa của '{word}': {set(antonyms)}")

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


Nhập vào từ khóa: clean
Từ đồng nghĩa của 'clean': {'fresh', 'unobjectionable', 'strip', 'sporting', 'white', 'make_clean', 'uncontaminating', 'plum', 'uninfected', 'sportsmanlike', 'unclouded', 'houseclean', 'cleanse', 'clean', 'blank', 'clear', 'scavenge', 'clean-living', 'plumb', 'pick', 'sporty', 'neat', 'clean_and_jerk', 'fairly', 'fair', 'clean_house', 'light'}
Từ trái nghĩa của 'clean': {'dirty', 'unfairly', 'unclean'}


8. Viết chương trình Python với thư viện NLTK để có cái nhìn tổng quan về bộ tag, chi tiết của
một tag cụ thể trong bộ tag và chi tiết về một số bộ tag liên quan, sử dụng biểu thức chính
quy.

In [80]:
import nltk
import re

# Tải xuống dữ liệu NLTK cần thiết
nltk.download('tagsets')

# Tổng quan về bộ tag
print("Tổng quan về các bộ tag:")
nltk.help.upenn_tagset()

# Chi tiết của một tag cụ thể
tag = input("Nhập vào tag bạn muốn xem chi tiết: ")
print(nltk.help.upenn_tagset(tag))

# Chi tiết về một số bộ tag liên quan (sử dụng biểu thức chính quy)
pattern = input("Nhập vào biểu thức chính quy để tìm các tag liên quan: ")
print("Các tag liên quan:")
for tag in nltk.data.load('help/tagsets/upenn_tagset.pickle').keys():
    if re.search(pattern, tag):
        print(f"- {tag}: {nltk.help.upenn_tagset(tag)}")

[nltk_data] Downloading package tagsets to /root/nltk_data...
[nltk_data]   Package tagsets is already up-to-date!


Tổng quan về các bộ tag:
$: dollar
    $ -$ --$ A$ C$ HK$ M$ NZ$ S$ U.S.$ US$
'': closing quotation mark
    ' ''
(: opening parenthesis
    ( [ {
): closing parenthesis
    ) ] }
,: comma
    ,
--: dash
    --
.: sentence terminator
    . ! ?
:: colon or ellipsis
    : ; ...
CC: conjunction, coordinating
    & 'n and both but either et for less minus neither nor or plus so
    therefore times v. versus vs. whether yet
CD: numeral, cardinal
    mid-1890 nine-thirty forty-two one-tenth ten million 0.5 one forty-
    seven 1987 twenty '79 zero two 78-degrees eighty-four IX '60s .025
    fifteen 271,124 dozen quintillion DM2,000 ...
DT: determiner
    all an another any both del each either every half la many much nary
    neither no some such that the them these this those
EX: existential there
    there
FW: foreign word
    gemeinschaft hund ich jeux habeas Haementeria Herr K'ang-si vous
    lutihaw alai je jour objets salutaris fille quibusdam pas trop Monte
    terram fiche oui corpor

9. Viết chương trình Python với thư viện NLTK để so sánh sự giống nhau của hai danh từ đã
cho.

In [81]:
nltk.download('wordnet')

def compare_noun_similarity(noun1, noun2):
  synset1 = wn.synsets(noun1, pos=wn.NOUN)
  synset2 = wn.synsets(noun2, pos=wn.NOUN)

  if not synset1 or not synset2:
    return 0

  max_similarity = 0
  for s1 in synset1:
    for s2 in synset2:
      similarity = s1.path_similarity(s2)
      if similarity is not None and similarity > max_similarity:
        max_similarity = similarity

  return max_similarity

if __name__ == "__main__":
  noun1 = input("Nhập danh từ thứ nhất: ")
  noun2 = input("Nhập danh từ thứ hai: ")

  similarity = compare_noun_similarity(noun1, noun2)
  print(f"Độ giống nhau giữa '{noun1}' và '{noun2}' là: {similarity}")


[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


Nhập danh từ thứ nhất: Apple
Nhập danh từ thứ hai: Car
Độ giống nhau giữa 'Apple' và 'Car' là: 0.07692307692307693


10. Viết chương trình Python với thư viện NLTK để so sánh sự giống nhau của hai động từ đã
cho.


In [82]:
def compare_verb_similarity(verb1, verb2):
  synset1 = wn.synsets(verb1, pos=wn.VERB)
  synset2 = wn.synsets(verb2, pos=wn.VERB)

  if not synset1 or not synset2:
    return 0

  max_similarity = 0
  for s1 in synset1:
    for s2 in synset2:
      similarity = s1.path_similarity(s2)
      if similarity is not None and similarity > max_similarity:
        max_similarity = similarity

  return max_similarity

if __name__ == "__main__":
  verb1 = input("Nhập động từ thứ nhất: ")
  verb2 = input("Nhập động từ thứ hai: ")

  similarity = compare_verb_similarity(verb1, verb2)
  print(f"Độ giống nhau giữa '{verb1}' và '{verb2}' là: {similarity}")


Nhập động từ thứ nhất: Run
Nhập động từ thứ hai: Jump
Độ giống nhau giữa 'Run' và 'Jump' là: 0.3333333333333333


11. Viết chương trình Python với thư viện NLTK để tìm số lượng tên nam và nữ trong các tên
kho ngữ liệu. In tên 10 nam và nữ đầu tiên. Lưu ý: Kho văn bản tên chứa tổng cộng khoảng
2943 nam (male.txt) và 5001 nữ (Female.txt) tên. Kho được biên soạn bởi Kantrowitz, Ross

In [84]:
# Tải xuống dữ liệu NLTK cần thiết
nltk.download('names')

# Đọc tên nam và nữ từ kho ngữ liệu
male_names = nltk.corpus.names.words('male.txt')
female_names = nltk.corpus.names.words('female.txt')

# Số lượng tên nam và nữ
num_male_names = len(male_names)
num_female_names = len(female_names)

print(f"Số lượng tên nam: {num_male_names}")
print(f"Số lượng tên nữ: {num_female_names}")

print("\n10 tên nam đầu tiên:")
for name in male_names[:10]:
    print(name)

print("\n10 tên nữ đầu tiên:")
for name in female_names[:10]:
    print(name)

Số lượng tên nam: 2943
Số lượng tên nữ: 5001

10 tên nam đầu tiên:
Aamir
Aaron
Abbey
Abbie
Abbot
Abbott
Abby
Abdel
Abdul
Abdulkarim

10 tên nữ đầu tiên:
Abagael
Abagail
Abbe
Abbey
Abbi
Abbie
Abby
Abigael
Abigail
Abigale


[nltk_data] Downloading package names to /root/nltk_data...
[nltk_data]   Package names is already up-to-date!


12. Viết chương trình Python với thư viện NLTK để in 15 kết hợp ngẫu nhiên đầu tiên được gắn
nhãn nam và được gắn nhãn tên nữ từ kho tên

In [85]:
# Đọc tên nam và nữ từ kho ngữ liệu
male_names = nltk.corpus.names.words('male.txt')
female_names = nltk.corpus.names.words('female.txt')

# Tạo danh sách các cặp tên (nam, nữ)
name_pairs = [(male, female) for male in male_names for female in female_names]

# Trộn ngẫu nhiên danh sách các cặp tên
random.shuffle(name_pairs)

# In 15 cặp tên ngẫu nhiên đầu tiên
print("15 cặp tên ngẫu nhiên (nam, nữ):")
for i in range(15):
    male, female = name_pairs[i]
    print(f"- {male}, {female}")

15 cặp tên ngẫu nhiên (nam, nữ):
- Tab, Luciana
- Osgood, Teri
- Caldwell, Prue
- Marlow, Cristen
- Hazel, Danielle
- Elliott, Rebe
- Alberto, Tillie
- Germaine, Shirlene
- Adnan, Malorie
- Tedmund, Adelle
- Spenser, Sybille
- Huntington, Sarena
- Darrin, Lavina
- Antin, Moyna
- Andres, Candide


13. Viết chương trình Python với thư viện NLTK để trích xuất ký tự cuối cùng của tất cả các tên
được gắn nhãn và tạo mảng mới với chữ cái cuối cùng của mỗi tên và nhãn được liên kết.

In [86]:
# Đọc tên nam và nữ từ kho ngữ liệu
male_names = nltk.corpus.names.words('male.txt')
female_names = nltk.corpus.names.words('female.txt')

# Tạo mảng mới với chữ cái cuối cùng của mỗi tên và nhãn được liên kết
labeled_last_letters = []
for name in male_names:
    labeled_last_letters.append((name[-1], 'male'))
for name in female_names:
    labeled_last_letters.append((name[-1], 'female'))

# In mảng mới
print("Mảng mới với chữ cái cuối cùng của mỗi tên và nhãn được liên kết:")
for last_letter, label in labeled_last_letters:
    print(f"- {last_letter}: {label}")

Kết quả truyền trực tuyến bị cắt bớt đến 5000 dòng cuối.
- l: female
- e: female
- y: female
- i: female
- e: female
- y: female
- l: female
- l: female
- e: female
- a: female
- a: female
- a: female
- h: female
- e: female
- a: female
- e: female
- s: female
- l: female
- a: female
- e: female
- e: female
- e: female
- a: female
- d: female
- e: female
- a: female
- e: female
- a: female
- y: female
- i: female
- a: female
- a: female
- a: female
- e: female
- e: female
- e: female
- a: female
- a: female
- s: female
- n: female
- a: female
- e: female
- a: female
- e: female
- n: female
- a: female
- e: female
- l: female
- a: female
- l: female
- g: female
- e: female
- a: female
- a: female
- e: female
- i: female
- e: female
- y: female
- a: female
- a: female
- s: female
- e: female
- e: female
- a: female
- a: female
- a: female
- a: female
- n: female
- s: female
- a: female
- e: female
- e: female
- n: female
- e: female
- y: female
- i: female
- a: female
- n: female
- e: fe

In [88]:
# Hàm để trích xuất ký tự cuối cùng và tạo mảng mới
def extract_last_character_with_label(name_list, label):
    last_characters = []
    for name in name_list:
        if name:
            last_char = name[-1]  # Lấy ký tự cuối cùng của tên
            last_characters.append((last_char, label))  # Thêm vào mảng với nhãn tương ứng
    return last_characters

# Trích xuất ký tự cuối cùng và nhãn từ danh sách các tên
male_last_chars = extract_last_character_with_label(male_names, 'male')
female_last_chars = extract_last_character_with_label(female_names, 'female')

# Kết hợp các mảng
all_last_chars = male_last_chars + female_last_chars

# In kết quả
print("Ký tự cuối cùng và nhãn tương ứng:")
for char, label in all_last_chars:
    print(f"{char} - {label}")

Kết quả truyền trực tuyến bị cắt bớt đến 5000 dòng cuối.
l - female
e - female
y - female
i - female
e - female
y - female
l - female
l - female
e - female
a - female
a - female
a - female
h - female
e - female
a - female
e - female
s - female
l - female
a - female
e - female
e - female
e - female
a - female
d - female
e - female
a - female
e - female
a - female
y - female
i - female
a - female
a - female
a - female
e - female
e - female
e - female
a - female
a - female
s - female
n - female
a - female
e - female
a - female
e - female
n - female
a - female
e - female
l - female
a - female
l - female
g - female
e - female
a - female
a - female
e - female
i - female
e - female
y - female
a - female
a - female
s - female
e - female
e - female
a - female
a - female
a - female
a - female
n - female
s - female
a - female
e - female
e - female
n - female
e - female
y - female
i - female
a - female
n - female
e - female
e - female
l - female
a - female
a - female
e - female
y - female
e - fema